File: Beginner/Java/DynamicProgramming/MaxSubArray.java
class Solution {
    public int maxSubArray(int[] nums) {
        // Kadane's algorithm
        // int maxEndHere = nums[0], maxSoFar = nums[0];
        // for (int i = 1; i < nums.length; i ++) {
        //     maxEndHere = Math.max(nums[i], nums[i] + maxEndHere);
        //     maxSoFar = Math.max(maxEndHere, maxSoFar);
        // }
        // return maxSoFar;
        
        //         divide and conquer
        return maxSub(nums, 0, nums.length - 1);
    }
    public int maxMid(int[] nums, int start, int mid, int end) {
        int leftSum = Integer.MIN_VALUE, rightSum = Integer.MIN_VALUE, midSum = 0;
        for (int i = mid; i >= start; i --) {
            midSum += nums[i];
            leftSum = Math.max(midSum, leftSum);
        }
        midSum = 0;
        for (int i = mid + 1; i <= end; i ++) {
            midSum += nums[i];
            rightSum = Math.max(midSum, rightSum);
        }
        return leftSum + rightSum;
    }
    
    public int maxSub(int[] nums, int start, int end) {
        if (start == end)
            return nums[start];
        int mid = (start + end) / 2;
        return Math.max(Math.max(maxSub(nums, start, mid), maxSub(nums, mid + 1, end)),
                        maxMid(nums, start, mid, end));
    }
}

File: Beginner/Python/DynamicProgramming/MaxSubArray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
            :type nums: List[int]
            :rtype: int
            """
        # Kadane's algorithm
        # max_ending_here = max_so_far = nums[0]
        # for x in nums[1:]:
        #     max_ending_here = max(x, max_ending_here + x)
        #     max_so_far = max(max_so_far, max_ending_here)
        # return max_so_far
        
        # divide and conquer
        return self.maxSubArraySum(nums, 0, len(nums)-1)
    
    def maxCrossingSum(self,arr, l, m, h) :
        sm = 0; left_sum = float("-inf")
        for i in range(m, l-1, -1) :
            sm = sm + arr[i]
            
            if (sm > left_sum) :
                left_sum = sm
        
        sm = 0; right_sum = float("-inf")
        for i in range(m + 1, h + 1) :
            sm = sm + arr[i]
            
            if (sm > right_sum) :
                right_sum = sm
        return left_sum + right_sum;

        def maxSubArraySum(self,arr, l, h) :
        if (l == h) :
            return arr[l]
        
        m = (l + h) // 2

        return max(self.maxSubArraySum(arr, l, m),
                   self.maxSubArraySum(arr, m+1, h),
                   self.maxCrossingSum(arr, l, m, h))
